### DISCUSSION TITLE
Use URLs in config where an endpoint is at one, not separate host/port/etc

### DISCUSSION THREAD

**Discussion by scop** [home-assistant/architecture#546] [General]
A bunch of integrations are using separate config and other parameters for host/port/protocol/ssl/path, and construct URLs they eventually use to reach an API endpoint out of those. This has at least two significant problems:

- It is not user friendly, easier would be to just input the URL, which is likely also copy-pasteable from somewhere.
- All of those parameters (and perhaps even more in some cases) are needed in order to be able to construct the URL in a way that it applies to all setups. For example, the services could reside in a custom URL path, custom port, https vs http, etc etc, for example if they're reverse proxied for whatever reason, stunneled, or something. The exhaustive set of things needing configuration this way might even be all components of an URL.

A random example is in the [dd-wrt device tracker](https://github.com/home-assistant/core/blob/7f29d028a35b6dd543b6a367aaf03df6fc80aebe/homeassistant/components/ddwrt/device_tracker.py#L57) -- just a sample, this pattern is widely spread in the tree I think.

This particular implementation is not appropriately configurable if the service is for whatever reason on a custom port or path (well unless one abuses the host parameter and concatenates those to it, but that's kind of a bug made possible by insufficient validation of the host config parameter, so let's not consider this an appropriate workaround -- I'm sure there are cases in the tree where such abuses aren't doable). Instead of configuring host and protocol, we should just configure the complete (base) URL to the service.

What I think is a good implementation is how I've made it [in huawei_lte](https://github.com/home-assistant/core/blob/3a0b0380c7acbbe06f793c8b11bdc962609f0002/homeassistant/components/huawei_lte/config_flow.py#L111). There, we take the complete base URL to the API as the config parameter. Additionally, we pass it through [`url_normalize`](https://github.com/niksite/url-normalize/blob/master/url_normalize/url_normalize.py#L206) which will do cleanups and normalization. These are good to do in the first place in my opinion, but in huawei_lte it is also important from the point of view that the URLs are (for better or for worse) used as keys to distinguish between routers in dicts etc. Another nice thing about `url_normalize` is that it can cleanly add a default scheme if the user entered just an IP address or hostname. This makes it even more user friendly. A downside to this is that the normalization/validation is not done clientside, but I think that's quite acceptable in this case.

I suggest we start to require API endpoints configured as URLs, not separate components of them, and move towards migrating existing integrations and their configs towards that as well.

One consideration to specify is to what kind of URLs we should apply this to. I think http, https, and ftp are quite clearly such ones, but for example telnet (ick) is one that I think is more commonly represented as host+port. I would probably extend this to all kinds of things representable as URLs, but don't have that strong opinion about it as long as at least http, https, and ftp are done with them.

**Comment by frenck** [comment]
IMHO this really depends on the use case.
Especially when using default, non-standard HTTP(s) ports, the URL doesn't make it easier to understand.

**Reply by scop** [reply]
I think I'm in disagreement with that, but just to make sure I understand what you're saying, by "default, non-standard" you mean not the default port of http (80) or https (443), but rather a default port in which some particular endpoint for a service/integration usually is at, if it's not the 80 for http or 443 for https?

**Reply by frenck** [reply]
Yes, for example, Octoprint uses port 5000 by default: `https://192.168.1.2:5000`

This is now unclear IMHO. Many users don't realize the `:5000` is actually the port. What is even more "tickling" in this particular example, is: Do I need to specify the API path in that same URL? Does it need to end with a slash? Username password as part of the URL? Or in this service, it isn't needed?

Additionally, the HTTPS in this case still needs a separate checkbox to ignore cert checks (so that isn't solved by just the URL).

And this is just OctoPrint as an example, the same applies to most of the integrations we actually provide.

One advantage of the separate fields is that we can now (since we introduced that recently) hide fields like ports (with their defaults) when advanced mode in the user profile is turned off (and thus just ask for the host for those users, nothing more).

We can provide defaults for e.g., a port, which cannot be done with the URL-based input method.

Lastly, when discovering partial data via discovery, we can hide or pre-fill certain fields during setup, while that is not possible with an URL-based input method.

**Reply by scop** [reply]
Yeah, I agree that if the access configuration for something typically isn't an URL, or its docs use something else, we shouldn't probably forcefully make it so. Not familiar with octoprint so can't say if that's the case for it, but sounds plausible. You also know the whole codebase better than I do, so no point in arguing against "most integrations" but I have come across a bunch for which IMO there's just no reason not to use an URL.

The rest of your points sound valid as well, so I'm becoming somewhat "softer" with my opinion on this, and agree with that it depends on the case. I do still maintain that there is quite a few places where the URL would be a better choice.

But consistency is also very important, so it would be good to have as few general approaches to this as possible. Another one if configuring in parts for catering to more advanced scenarios would be to consistently make scheme, host, port, and path configurable in advanced mode in everything where we construct an URL out of them under the hood. Path config/validation could cater for possible query string in the config, and host for embedding username:password@ in it (and integrations prepare to behave sensibly when they're there).
